Update build to optionally configure with nmake

adiaaida opened the pull request:
For formatting, we want to be able to obtain the compile_commands.json
file so that we do not have to 1) do a full build of coreclr and 2)
generate it ourselves from the build log. This change modifies build.cmd
to take an option "usenmakemakefiles." This option will set
__NMakeMakefiles to 1, which is used by the gen-buildsys-win.bat script.
This option also sets all the same options as configureonly, as we cannot
do a full build using NMake Makefiles, and we only want to use this to
force CMake to generate a compile_commands.json file for the jit
directory, which it can do using NMake Makefiles, but not Visual Studio.
The update to gen-buildsys-win.bat checks the value of __NMakeMakefiles,
and if it is set to 1, it sets the CMake Generator to NMake Makefiles.

adiaaida commented:
@RussKeldorph 
/cc: @dotnet/jit-contrib

jashook commented:
@adiaaida does this generate the correct compile commands for arm64? The cmake generator for arm64 was hacked around because there is no arm64 windows CMake generator support yet. It is very possible the compile commands are changed from what the x64 CMake VS Generator gives.

adiaaida commented:
I don't know. I've only tested it with x64, since that is the only architecture jit-format currently supports.

RussKeldorph commented:
Are we still requiring that you build before formatting?  If so, is it reasonable to remove that limitation...i.e. just have jit-format generate compile-commands by itself?  Is that a quick process?

RussKeldorph commented:
LGTM

adiaaida commented:
You won't have to do a full build, but you'll always have to have something generate compile_commands.json, whether that be by running build.cmd usenmakemakefiles or doing a full build and having jit-format generate compile_commands.json from the build log. Or are you suggesting that jit-format could call build.cmd usenmakemakefiles? The problem here is that cmake gets unhappy when you have a CMakeCache.txt or cmake.definitions file in your bin\obj\<build> directory where the generator doesn't match the generator you are currently trying to use (unless there's a way to tell build.cmd to build into a different directory?), so I think people are going to have to manually run that step so they can delete the files from obj that need to be deleted.

RussKeldorph commented:
My preference would be for jit-format to "just work" whether the build has run or not.  Similar to how clang-format is build agnostic.  If that means plumbing a way to redirect CMake output files to a jit-format-specific temp location to avoid conflicts, that seems like a small price to pay.  If that's possible, how long would it take to implement?

adiaaida commented:
If it's possible, maybe an hour, but figuring out if it's possible might take a bit longer.

adiaaida commented:
@RussKeldorph I've modified this change to build NMake runs to a different dir called "nmakeobj." In conjunction, I've updated my jitutils PR (https://github.com/dotnet/jitutils/pull/30) to look for the compile_commands.json in that dir, and if it doesn't exist, to run build.cmd <arch> <build> usenmakemakefiles.

RussKeldorph commented:
LGTM
